### PR TITLE
New benchmarks

### PR DESCRIPTION
**Pull request by tiagolascasas**
Adds four new benchmark sets:

- Rosetta (5 applications)
- CHStone (12 kernels)
- HiFlipVX (one library and an example application that uses said library)
- Parboil (11 applications)

All benchmarks have been verified in terms of parsing, for the specified language standard (C90 or C++11, depending on the source code). Some required extensive restructuring of the .c and .h files. These benchmarks only have a single input size. Rosetta needed an additional flag, -D SW, to tell the preprocessor to consider the software version of the application, and ignore the OpenMP/HLS options.

Compilation and execution have not been tested yet. These benchmarks are for targeting HLS, and as such it is not immediately relevant to execute them after being modified by Clava. Still, we do not anticipate any major issues about compilation, as they do not require any external libraries or flags other than those already configured (such as -lm).

Test the parsing with this code, after including the benchmarks in the Clava config file:

```
import lara.benchmark.PolybenchBenchmarkSet;
import lara.benchmark.HiFlipVXBenchmarkSet;
import lara.benchmark.CHStoneBenchmarkSet;
import lara.benchmark.RosettaBenchmarkSet; 
import lara.benchmark.ParboilBenchmarkSet; 
import weaver.WeaverJps;

aspectdef Main
	var b;

	b = new PolybenchBenchmarkSet();
	printBench(b, ["adi", "deriche", "lu"], "SMALL");

	b = new HiFlipVXBenchmarkSet();
	printBench(b, "v2", "N");

	b = new CHStoneBenchmarkSet();
	var names = [
		"adpcm", 		
		"aes", 		
		"blowfish", 	
		"dfadd", 		
		"dfdiv", 		
		"dfmul", 		
		"dfsin", 		
		"gsm", 		
		"jpeg", 		
		"mips", 		
		"motion", 	
		"sha"		
	];
	printBench(b, names, "N");

	b = new RosettaBenchmarkSet();
	var names = [
		"3d-rendering", 		
		"digit-recognition", 	
		"face-detection", 		
		"optical-flow", 		
		"spam-filter", 		
	];
	printBench(b, names, "N");

	b = new ParboilBenchmarkSet();
	var names = [
		"bfs", 			
		"cutcp", 			
		"histo",			
		"lbm",		
		"mri-gridding", 	
		"mri-q", 			
		"sad", 			
		"sgemm", 			
		"spmv", 			
		"stencil", 		
		"tpacf"			
	];
	printBench(b, names, "N");
end

// Prints the name of all functions in the bench, as a sanity check
function printBench(benchmarks, names, sizes) {
	benchmarks.setBenchmarks(names);
	benchmarks.setInputSizes(sizes);
	
	var res = [];
	for (var b of benchmarks) {
		b.load();
		var f = [];
		for (var elem of WeaverJps.search("function")) {
			if (elem.isImplementation) {
				f.push(elem.name);
			}
		}
		res.push(f.join(","));
	}
	println(res.join("\n"));
}
```

### CODE REVIEW

File: CHStone/resources/adpcm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+add_benchmark_cosimulation_test("adpcm_main")
+add_benchmark_synthesis_test("adpcm_main")
+add_hls_target("adpcm_main")
```
Comment (joaobispo) [inline CHStone/resources/adpcm/CMakeLists.txt:3]: Maybe remove these CMakeLists.txt files, if they are not needed

File: HiFlipVX/lara/benchmark/HiFlipVXBenchmarkInstance.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+laraImport("clava.Clava");
+laraImport("clava.ClavaJoinPoints");
+
+laraImport("weaver.WeaverJps");
+
+/**
+ * Instance of a C++ HiFlipVX benchmark.
+ */
+function HiFlipVXBenchmarkInstance(benchmarkName, inputSize) {
```
Comment (joaobispo) [inline HiFlipVX/lara/benchmark/HiFlipVXBenchmarkInstance.js:12]: When running HiFlipVX, there is the following message:

`Cannot open file to read!: No such file or directory`

You can customize how an application runs by overloading the function [_executePrivate() in BenchmarkInstance](https://github.com/specs-feup/lara-framework/blob/master/LaraApi/src-lara/lara/benchmark/BenchmarkInstance.lara)

**Review by joaobispo — COMMENTED** [review]
These benchmarks have not been tested for compilation or running, in some cases the code does not correctly compile (e.g. Parboil lbm), in others it is missing inputs for execution (HiFlipVX, Parboil). Even in the HLS case, after transformations you will want to verify if the transformation did not produce incorrect code, it is important to be able to generate code that executes locally.

Please test compilation and execution with code as the following:

```
// Example for Parboil
laraImport("lara.benchmark.ParboilBenchmarkSet")

// In case you need to customize how CMake must be called
function windowsCMaker() {
	const cmaker = new CMaker();
	cmaker.setGenerator("MinGW Makefiles");	
	cmaker.setMakeCommand("mingw32-make");

	return cmaker;
}

// Enable debug messages
setDebug(true)

var set = new ParboilBenchmarkSet();
set.setBenchmarks("lbm"); // If there are errors, it is easier to go one by one

// Set custom CMake options
set.setCMakerProvider(windowsCMaker)

for(var instance of set) {
	println("Benchmark : " + instance.getName())
	
	instance.execute()
}
```

File: Parboil/resources/lbm/main.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include "config.h"
+#include "parboil.h"
+
+/*############################################################################*/
+
+typedef struct
+{
+	float timeScale;
+	clock_t tickStart, tickStop;
+	struct tms timeStart, timeStop;
```
Comment (joaobispo) [inline Parboil/resources/lbm/main.h:25]: `struct tms` is most likely a bug, include `time.h` [declares a struct tm but not tms](https://pubs.opengroup.org/onlinepubs/007908799/xsh/time.h.html).

**Comment by tiagolascasas** [comment]
All benchmarks are now successfully compiling and executing, at least on Windows.

The single exception is Rosetta's optical-flow, due to [this issue](https://github.com/specs-feup/clava/issues/106)

Note that every benchmark suite has been [documented here](https://github.com/specs-feup/clava/wiki/Benchmarks), which includes information about the possible input sizes for each app/kernel.

As a note, some input files exceed 50MB, and git recommends us to enable LFS in the repository.

**Pull request merged**